pkg/agent/openflow/pod_connectivity.go

```diff
@@ -157,16 +163,105 @@ func (f *featurePodConnectivity) initFlows() []binding.Flow {
 		flows = append(flows, getCachedFlows(cachedFlows)...)
 	}
 
 	return flows
 }
+
+// trafficControlMarkFlows generates the flows to mark the packets that need to be redirected or mirrored with the value of
+// the destination output port.
+func (f *featurePodConnectivity) trafficControlMarkFlows(srcOfPorts []uint32, dstOfPort uint32, direction v1alpha2.Direction) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	for _, port := range srcOfPorts {
+		if direction == v1alpha2.DirectionIn || direction == v1alpha2.DirectionBoth {
+			flows = append(flows, IngressTrafficControlMarkTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchRegFieldWithValue(TargetOFPortField, port).
+				Action().LoadToRegField(TrafficControlOFPortField, dstOfPort).
+				Action().NextTable().
+				Done())
+		}
+		if direction == v1alpha2.DirectionOut || direction == v1alpha2.DirectionBoth {
+			flows = append(flows, EgressTrafficControlMarkTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchInPort(port).
+				Action().LoadToRegField(TrafficControlOFPortField, dstOfPort).
+				Action().NextTable().
+				Done())
+		}
+	}
+	return flows
+}
+
+// trafficControlInitFlows generates the initial flows for the traffic control mark value.
+func (f *featurePodConnectivity) trafficControlInitFlows(ofInPort, ofOutPort uint32, action v1alpha2.Action) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	if action == v1alpha2.ActionMirror {
+		// This generates the flow to output packets to the original out port as well as mirror the packets to the ofInPort
+		// which is used to receive mirrored packets in L2ForwardingOutTable.
+		flows = append(flows, L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh+1).
+			Cookie(cookieID).
+			MatchRegFieldWithValue(TrafficControlOFPortField, ofInPort).
+			MatchRegMark(OFPortFoundRegMark).
+			Action().OutputToRegField(TargetOFPortField).
+			Action().Output(ofInPort).
+			Done())
+	} else if action == v1alpha2.ActionRedirect {
+		flows = append(flows,
+			// This generates the flow to forward the redirected packets sourced from the ofOutPort to stageRouting directly
+			// in ClassifierTable. Note that, for the packets which are to be originally output to a tunnel, value of
+			// NXM_NX_TUN_IPV4_DST for the redirected packets needs to be loaded in stageRouting.
+			ClassifierTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchInPort(ofOutPort).
+				Action().LoadRegMark(FromRedirectRegMark).
+				Action().GotoStage(stageRouting).
+				Done(),
+			// This generates the flow to forward the packets to be redirected to the ofInPort which is used to receive
+			// the packets.
+			L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh+1).
+				Cookie(cookieID).
+				MatchRegFieldWithValue(TrafficControlOFPortField, ofInPort).
```

Comment:
Does it mean a port can only be either target port of mirror action or redirect action but not both? otherwise there would be a conflict.

Reply:
Yes, in current code. Should a target port support both actions?

Reply:
Updated with supporting both actions.

Reply:
I think it should be supported, otherwise some scenarios that have both IDS and IPS will need multiple instances. What's the problem of supporting both actions?

Reply:
Some ideas:
- If a target port only supports action mirror, we don't need the flows to forward the returned packets for action redirect. I think the capability of a target port (e.g. only supporting action mirror; only supporting action mirror; supporting both) should be passed to `InstallTrafficControlCommonFlows`.
- If a target port only supports action mirror, it cannot be the target port of traffic control rules whose action is redirection.
- If a target port supports action mirror and redirection, if a traffic control rule whose action is mirror is installed, then for IDS/IPS, its the IDS/IPS responsibility to filter the mirrored packets from redirected packets. If IDS/IPS cannot do this, and the mirrored packets return OVS pipeline again, I think we cannot filter it out.